### ISSUE TITLE
Move XamlDocumentDiagnosticAnalyzer to OOP

### ISSUE THREAD

**Issue by tmat** [dotnet/roslyn#42183]
It currently depends on XAML-specific TextBuffer [properties](https://devdiv.visualstudio.com/DevDiv/_git/VS?path=%2Fsrc%2FXaml%2FDesigner%2FSource%2FXamlLanguageService%2FCodeAnalysis%2FDiagnostics%2FAnalyzers%2FXamlDocumentAnalyzerService.cs&version=GBmaster&line=123&lineEnd=124&lineStartColumn=1&lineEndColumn=1&lineStyle=plain). 

A short term workaround would be to move the analyzer OOP and let it call back to the in-proc XAML service, passing it an in-proc TextBuffer.

**Comment by davidwengier** [comment]
Self-notes:

* Currently, in-proc solution crawler runs `XamlDocumentDiagnosticAnalyzer`, which calls in to `XamlDocumentAnalyzerService` on the Xaml side to do the work (linked above)
* To remove in-proc solution crawler, we need `XamlDocumentDiagnosticAnalyzer` to be run by the OOP solution crawler
* Since `XamlDocumentAnalyzerService` pulls things out of the text buffer property bag, it can't be OOP
* Solution is to create an indirection layer so OOP solution crawler runs `XamlDocumentDiagnosticAnalyzer`, which calls a new service running in-proc, which calls `XamlDocumentAnalyzerService` on the Xaml side to do the work

Worth checking with Xaml team first what this does, what its for, what are future plans. Long term we'd like it to go to pull model and not depend on solution crawler at all.

**Comment by tmat** [comment]
@mgoertz-msft IS working on XAML LSP. I believe there are some work items still open on our side to allow XAML transition away from this analyzer.

**Comment by mgoertz-msft** [comment]
Correct, we are not quite ready to move OOP because we still have a lot of dependencies on Visual Studio, the least of which is the ITextBuffer dependency.

Somewhat orthogonal to this is that we cannot move off of DocumentAnalyzer either because there is no way to register code fixes/refactorings for TextDocuments at this point. They all require real Documents, which is what we are currently providing from XAML projects that shadow the managed projects.

**Comment by davidwengier** [comment]
New approach discussed with Sam, because DocumentDiagnosticAnalyzers are not supported in OOP right now, is to move our internal hook up to ISB, so that when the DocumentDiagnosticAnalyzer infrastructure moves to OOP, this will continue to work with just a change to the service registration.

This will require a change on the Xaml side to take DocumentIds instead of Documents.

**Comment by mgoertz-msft** [comment]
@davidwengier Are you saying that AnalyzeSyntaxAsync(Document document, CancellationToken cancellationToken) will switch to AnalyzeSyntaxAsync(DocumentId documentId, CancellationToken cancellationToken) and we will need to ask the CurrentSolution  to give us the Document for that DocumentId?

That sounds fine. On the other hand, you could do that before calling AnalyzeSyntaxAsync() too and keep the API the same. ;-)

**Comment by davidwengier** [comment]
Yeah, whether "the Xaml side" is the analyzer in the Roslyn repo or the analyzer service in the VS repo, shouldn't matter at all.

**Comment by davidwengier** [comment]
This isn't possible at the moment, the DocumentDiagnosticAnalyzer cannot be moved OOP.
